### PR TITLE
Disarm damage/slow damage now closes solar

### LINKED ISSUE #3310: Disarm and Capture damage should close Solar
Apparently Racketeer and Thunderbird do not cause Solars to close. They should cause it to close. In general, any damage that is reduced by armour should cause Solar to close.

### PR DESCRIPTION
**Pull request by Arch-Shaman**
Fixes #3310.
* Adds pure disarm and slow to the solar's damage check.

### CODE REVIEW

**Comment by GoogleFrog** [comment]
See https://github.com/ZeroK-RTS/Zero-K/pull/4001

File: scripts/energysolar.lua [hunk 1]
```diff
@@ -116,7 +116,7 @@ end
```
Comment (sprunk) [inline scripts/energysolar.lua:116]:
```suggestion
local applyDefensiveManeuverDefs = {}
do
	local utils = Spring.Utilities
	local s = utils.IsPureSlowWeapon
	local c = utils.IsCaptureWeapon
	local d = utils.IsPureDisarmWeapon

	for i = 1, #WeaponDefs do
		applyDefensiveManeuverDefs[i] = s(i) or c(i) or d(i)
	end
end
```

File: scripts/energysolar.lua [hunk 2]
```diff
@@ -116,7 +116,7 @@ end
 
 -- this happens before PreDamaged but only in 97.0+
 function script.HitByWeapon(x, z, weaponDefID, damage)
-	if damage > 0 and not (weaponDefID and noFFWeaponDefs[weaponDefID]) then
+	if (damage > 0 or Spring.Utilities.IsPureSlowWeapon(weaponDefID) or Spring.Utilities.IsCaptureWeapon(weaponDefID) or Spring.Utilities.IsPureDisarmWeapon(weaponDefID)) and not (weaponDefID and noFFWeaponDefs[weaponDefID]) then
```
Comment (sprunk) [inline scripts/energysolar.lua:119]:
```suggestion
	if (damage > 0 or applyDefensiveManeuverDefs[weaponDefID]) and not (weaponDefID and noFFWeaponDefs[weaponDefID]) then
```

**Comment by GoogleFrog** [comment]
If we're being really finicky, then I'd localise `Spring.Utilities.IsPureSlowWeapon` since it could be called a lot. I disagree with Sprung's suggestion of duplicating a table the size of #WeaponDefs for every living Solar. noFFWeaponDefs could also be moved to this new utility file for special weapon types.

Also, `if bla then return true else return false end` can be replaced with `return (bla and true) or false`

**Comment by sprunk** [comment]
> I disagree with Sprung's suggestion of duplicating a table the size of #WeaponDefs for every living Solar.

In that case how about `Spring.Utilities.LegitZeroDamageWeapon` (possibly with a better name) which could also be used for the "status damage causes decloak" #4001 gadget?

**Comment by Arch-Shaman** [comment]
On it.

**Comment by Arch-Shaman** [comment]
The name probably isn't the world's best name, but it was the most descriptive thing I could come up with.

File: LuaRules/Utilities/SpecialWeaponCaching.lua [hunk 3]
```diff
@@ -15,25 +15,17 @@ for i = 1, #WeaponDefs do
 end
 
 function Spring.Utilities.IsPureSlowWeapon(weaponID)
-	if pureSlowWeapons[weaponID] then
-		return true
-	else
-		return false
-	end
+	return (pureSlowWeapons[weaponID] and true) or false
```
Comment (sprunk) [inline LuaRules/Utilities/SpecialWeaponCaching.lua:18]: how about just
```suggestion
	return pureSlowWeapons[weaponID]
```
(if `false` vs `nil` is a problem then set stuff to false in the table)

File: LuaRules/Utilities/SpecialWeaponCaching.lua [hunk 4]
```diff
@@ -15,25 +15,17 @@ for i = 1, #WeaponDefs do
...
 function Spring.Utilities.IsPureDisarmWeapon(weaponID)
-	if pureDisarmWeapons[weaponID] then
-		return true
-	else
-		return false
-	end
+	return (pureDisarmWeapons[weaponID] and true) or false
 end
+
+function Spring.Utilities.GetWeaponHasStatusEffect(weaponID) -- Single call for checking all three
```
Comment (sprunk) [inline LuaRules/Utilities/SpecialWeaponCaching.lua:29]: `IsWeaponPureStatusEffect`? Disruptors and the scout boat have status effects so sound like they should be counted but wouldn't as unpure.

File: LuaRules/Utilities/SpecialWeaponCaching.lua [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+for i = 1, #WeaponDefs do
+	if WeaponDefs[i].customParams.disarmdamageonly == '1' then
+		pureDisarmWeapons[i] = true
+	end
+	if WeaponDefs[i].customParams.timeslow_onlyslow == '1' then
+		pureSlowWeapons[i] = true
+	end
+	if WeaponDefs[i].customParams.is_capture == '1' then
+		captureWeapons[i] = true
+	end
```
Comment (sprunk) [inline LuaRules/Utilities/SpecialWeaponCaching.lua:14]:
```suggestion
	local cp = WeaponDefs[i].customParams
	pureDisarmWeapons[i] = (cp.disarmdamageonly  == '1')
	pureSlowWeapons  [i] = (cp.timeslow_onlyslow == '1')
	captureWeapons   [i] = (cp.is_capture        == '1')
```

File: LuaRules/Utilities/SpecialWeaponCaching.lua (hunk 4 above)
Comment (Arch-Shaman) [reply LuaRules/Utilities/SpecialWeaponCaching.lua:29]: That sounds better.

File: LuaRules/Utilities/SpecialWeaponCaching.lua (hunk 3 above)
Comment (Arch-Shaman) [reply LuaRules/Utilities/SpecialWeaponCaching.lua]: 1aae5a2

File: LuaRules/Utilities/SpecialWeaponCaching.lua (hunk 5 above)
Comment (Arch-Shaman) [reply LuaRules/Utilities/SpecialWeaponCaching.lua]: 1aae5a2

**Comment by Arch-Shaman** [comment]
#4001 update is getting pr'd after testing.

**Pull request merged**